src/oscar/apps/customer/utils.py

```diff
@@ -1,131 +1,136 @@
-            dispatched_messages = self.dispatch_anonymous_messages(email, messages)
-        else:
-            dispatched_messages = self.dispatch_user_messages(order.user, messages)
+    # Event codes
+    REGISTRATION_EVENT_CODE = 'REGISTRATION'
+    PASSWORD_RESET_EVENT_CODE = 'PASSWORD_RESET'
+    PASSWORD_CHANGED_EVENT_CODE = 'PASSWORD_CHANGED'
+    EMAIL_CHANGED_EVENT_CODE = 'EMAIL_CHANGED'
+    PRODUCT_ALERT_EVENT_CODE = 'PRODUCT_ALERT'
+    PRODUCT_ALERT_CONFIRMATION_EVENT_CODE = 'PRODUCT_ALERT_CONFIRMATION'
 
-        self.create_communication_event(order, event_type, dispatched_messages)
+    def __init__(self, logger=None, mail_connection=None):
+        self.dispatcher = Dispatcher(logger=logger, mail_connection=mail_connection)
 
-    def dispatch_anonymous_messages(self, email, messages):
-        dispatched_messages = {}
-        if email:
-            dispatched_messages['email'] = self.send_email_messages(email, messages), None
-        return dispatched_messages
+    def send_registration_email_for_user(self, user, extra_context):
+        messages = self.dispatcher.get_messages(self.REGISTRATION_EVENT_CODE, extra_context)
+        self.dispatcher.dispatch_user_messages(user, messages)
 
-    def dispatch_user_messages(self, user, messages):
-        """
-        Send messages to a site user
-        """
-        dispatched_messages = {}
-        if messages['subject'] and (messages['body'] or messages['html']):
-            dispatched_messages['email'] = self.send_user_email_messages(user, messages)
-        if messages['sms']:
-            dispatched_messages['sms'] = self.send_text_message(user, messages['sms'])
-        return dispatched_messages
+    def send_password_reset_email_for_user(self, user, extra_context):
+        messages = self.dispatcher.get_messages(self.PASSWORD_RESET_EVENT_CODE, extra_context)
+        self.dispatcher.dispatch_user_messages(user, messages)
 
-    # Internal
+    def send_password_changed_email_for_user(self, user, extra_context):
+        messages = self.dispatcher.get_messages(self.PASSWORD_CHANGED_EVENT_CODE, extra_context)
+        self.dispatcher.dispatch_user_messages(user, messages)
 
-    def create_communication_event(self, order, event_type, dispatched_messages):
-        """
-        Create order communications event for audit
-        """
-        if dispatched_messages and event_type is not None:
-            CommunicationEvent._default_manager.create(order=order, event_type=event_type)
+    def send_email_changed_email_for_user(self, user, extra_context):
+        messages = self.dispatcher.get_messages(
+            self.EMAIL_CHANGED_EVENT_CODE, extra_context)
+        self.dispatcher.dispatch_user_messages(user, messages)
 
-    def create_customer_email(self, user, messages, email):
-        """
-        Create Email instance in database for logging purposes.
+    def send_product_alert_email_for_user(self, product):  # noqa: C901 too complex
```

Comment:
This method is copied over from `customer.alerts.utils` but the original remains? I wonder whether for consistency, we should have a `CustomerAlertDispatcher` (or `AlertDispatcher`) for this method to live in?